File: Src/Icm.Core/Reflection/MethodInfoExtensions.vb
Imports System.Runtime.CompilerServices
Imports System.Reflection

Namespace Icm.Reflection

    Public Module MethodInfoExtensions

        <Extension()>
        Function GetAttributes(Of T As Attribute)(ByVal mi As MethodInfo, ByVal inherit As Boolean) As T()
            Return DirectCast(mi.GetCustomAttributes(GetType(T), inherit), T())
        End Function

        <Extension()>
        Function HasAttribute(Of T As Attribute)(ByVal mi As MethodInfo, ByVal inherit As Boolean) As Boolean
            Return DirectCast(mi.GetCustomAttributes(GetType(T), inherit), T()).Count > 0
        End Function

        <Extension()>
        Function GetAttribute(Of T As Attribute)(ByVal mi As MethodInfo, ByVal inherit As Boolean) As T
            Return DirectCast(mi.GetCustomAttributes(GetType(T), inherit), T()).Single
        End Function

    End Module

End Namespace

File: Src/Icm.Core/Reflection/MethodInfoExtensions.cs
using System;
using System.Linq;
using System.Reflection;

namespace Icm.Reflection
{
	public static class MethodInfoExtensions
	{
		/// <summary>
		/// Get the attributes of a given method.
		/// </summary>
		/// <typeparam name="T"></typeparam>
		/// <param name="mi"></param>
		/// <param name="inherit"></param>
		/// <returns></returns>
		/// <remarks></remarks>
		public static T[] GetAttributes<T>(this MethodInfo mi, bool inherit) where T : Attribute
		{
			return (T[])mi.GetCustomAttributes(typeof(T), inherit);
		}


		/// <summary>
		/// Do the method have the attribute?
		/// </summary>
		/// <typeparam name="T">Attribute type to check</typeparam>
		/// <param name="mi"></param>
		/// <param name="inherit"></param>
		/// <returns></returns>
		/// <remarks></remarks>
		public static bool HasAttribute<T>(this MethodInfo mi, bool inherit) where T : Attribute
		{
			return ((T[])mi.GetCustomAttributes(typeof(T), inherit)).Length > 0;
		}

		/// <summary>
		/// Get the given attribute, that must be unique.
		/// </summary>
		/// <typeparam name="T"></typeparam>
		/// <param name="mi"></param>
		/// <param name="inherit"></param>
		/// <returns></returns>
		/// <remarks></remarks>
		public static T GetAttribute<T>(this MethodInfo mi, bool inherit) where T : Attribute
		{
			return ((T[])mi.GetCustomAttributes(typeof(T), inherit)).Single();
		}
	}
}